lib/graphql/lang/ast_reducer.ex

```diff
@@ -0,0 +1,99 @@
+
+# Used for computing a result as a function of the an AST traversal.  The
+# traditional OO Visitor pattern does not really work well in functional
+# languages due to its reliance on generating side-effects.
+defmodule GraphQL.Lang.ASTReducer do
+
+  alias GraphQL.Lang.Visitor
+  alias GraphQL.Lang.VisitorContext
+  alias GraphQL.Lang.Stack
+
+  defmodule Nodes do
+    @kinds %{
+      Name:                       [],
+      Document:                   [:definitions],
+      OperationDefinition:        [:name, :variableDefinitions, :directives, :selectionSet],
+      VariableDefinition:         [:variable, :type, :defaultValue],
+      Variable:                   [:name],
+      SelectionSet:               [:selections],
+      Field:                      [:alias, :name, :arguments, :directives, :selectionSet],
+      Argument:                   [:name, :value],
+      FragmentSpread:             [:name, :directives],
+      InlineFragment:             [:typeCondition, :directives, :selectionSet],
+      FragmentDefinition:         [:name, :typeCondition, :directives, :selectionSet],
+      IntValue:                   [],
+      FloatValue:                 [],
+      StringValue:                [],
+      BooleanValue:               [],
+      EnumValue:                  [],
+      ListValue:                  [:values],
+      ObjectValue:                [:fields],
+      ObjectField:                [:name, :value],
+      Directive:                  [:name, :arguments],
+      NamedType:                  [:name],
+      ListType:                   [:type],
+      NonNullType:                [:type],
+      ObjectTypeDefinition:       [:name, :interfaces, :fields],
+      FieldDefinition:            [:name, :arguments, :type],
+      InputValueDefinition:       [:name, :type, :defaultValue],
+      InterfaceTypeDefinition:    [:name, :fields],
+      UnionTypeDefinition:        [:name, :types],
+      ScalarTypeDefinition:       [:name],
+      EnumTypeDefinition:         [:name, :values],
+      EnumValueDefinition:        [:name],
+      InputObjectTypeDefinition:  [:name, :fields],
+      TypeExtensionDefinition:    [:definition]
+    }
+
+    def kinds, do: @kinds
+  end
+
+  def reduce(node, visitor, acc) do
+    { _, acc } = visit(node, visitor, acc)
+    visitor |> Visitor.finish(acc)
+  end
+
+  defp visit(node, visitor, acc) do
+    visit(%VisitorContext{}, node, visitor, acc)
+  end
+
+  defp visit(context, [child|rest], visitor, acc) do
+    { next_action, acc } = visit(context, child, visitor, acc)
+    case next_action do
+      :continue -> visit(context, rest, visitor, acc)
```

Comment:
Calling visit recursively like this means I actually _will_ need to pop the stack in the context. Derp.